test/Microsoft.AspNetCore.Mvc.WebApiCompatShimTest/ApiControllerActionDiscoveryTest.cs

```diff
@@ -406,20 +405,49 @@ private void Invoke(ControllerActionDescriptorProvider provider, ActionDescripto
             provider.OnProvidersExecuted(context);
         }
 
-        private class NamespaceFilteredControllerTypeProvider : DefaultControllerTypeProvider
+        private static ApplicationPartManager GetApplicationManager(params TypeInfo[] controllerTypes)
         {
-            public NamespaceFilteredControllerTypeProvider(IAssemblyProvider provider)
-                : base(provider)
+            var manager = new ApplicationPartManager();
+            manager.ApplicationParts.Add(new TestPart(controllerTypes));
+            manager.FeatureProviders.Add(new TestProvider());
+            manager.FeatureProviders.Add(new NamespaceFilteredControllersFeatureProvider());
+            return manager;
+        }
+
+        private class TestPart : ApplicationPart, IApplicationPartTypeProvider
+        {
+            public TestPart(IEnumerable<TypeInfo> types)
             {
+                Types = types;
+            }
+
+            public override string Name => "Test";
+
+            public IEnumerable<TypeInfo> Types { get; }
+        }
 
+        private class TestProvider : IApplicationFeatureProvider<ControllerFeature>
+        {
+            public void PopulateFeature(IEnumerable<ApplicationPart> parts, ControllerFeature feature)
+            {
+                foreach (var type in parts.OfType<IApplicationPartTypeProvider>().SelectMany(t => t.Types))
+                {
+                    feature.Controllers.Add(type);
+                }
             }
+        }
 
-            public override IEnumerable<TypeInfo> ControllerTypes
+        private class NamespaceFilteredControllersFeatureProvider : IApplicationFeatureProvider<ControllerFeature>
```

Comment:
Keep this in your brain as an example we can point people to if they want to write a custom feature.